crud/tarantool_test.go

```diff
@@ -0,0 +1,601 @@
+			ObjectsOperationData(objectsOperationData).
+			Opts(opManyOpts),
+	},
+}
+
+func generateTuples() []crud.Tuple {
+	tpls := []crud.Tuple{}
+	for i := 1010; i < 1020; i++ {
+		tpls = append(tpls, crud.Tuple{uint(i), nil, "bla"})
+	}
+
+	return tpls
+}
+
+func generateTuplesOperationsData(tpls []crud.Tuple, operations []crud.Operation) []interface{} {
+	tuplesOperationsData := []interface{}{}
+	for _, tpl := range tpls {
+		tuplesOperationsData = append(tuplesOperationsData, []interface{}{
+			tpl,
+			operations,
+		})
+	}
+
+	return tuplesOperationsData
+}
+
+func generateObjects() []crud.Object {
+	objs := []crud.Object{}
+	for i := 1010; i < 1020; i++ {
+		objs = append(objs, crud.Object{
+			"id":   uint(i),
+			"name": "bla",
+		})
+	}
+
+	return objs
+}
+
+func generateObjectsOperationsData(objs []crud.Object, operations []crud.Operation) []interface{} {
+	objectsOperationsData := []interface{}{}
+	for _, obj := range objs {
+		objectsOperationsData = append(objectsOperationsData, []interface{}{
+			obj,
+			operations,
+		})
+	}
+
+	return objectsOperationsData
+}
+
+func getCrudError(req tarantool.Request, crudError interface{}) (interface{}, error) {
+	var err []interface{}
+	var ok bool
+
+	code := req.Code()
+	if crudError != nil {
+		if code == tarantool.Call17RequestCode {
+			return crudError, nil
+		}
+
+		if err, ok = crudError.([]interface{}); !ok {
+			return nil, fmt.Errorf("Incorrect CRUD error format")
+		}
+
+		if len(err) < 1 {
+			return nil, fmt.Errorf("Incorrect CRUD error format")
+		}
+
+		if err[0] != nil {
+			return err[0], nil
+		}
+	}
+
+	return nil, nil
+}
+
+func testCrudRequestPrepareData(t *testing.T, conn tarantool.Connector) {
+	t.Helper()
+
+	for i := 1010; i < 1020; i++ {
+		req := tarantool.NewReplaceRequest(spaceName).Tuple(
+			[]interface{}{uint(i), nil, "bla"})
+		if _, err := conn.Do(req).Get(); err != nil {
+			t.Fatalf("Unable to prepare tuples: %s", err)
+		}
+	}
+}
+
+func testSelectGeneratedData(t *testing.T, conn tarantool.Connector,
+	expectedTuplesCount int) {
+	req := tarantool.NewSelectRequest(spaceNo).
+		Index(indexNo).
+		Limit(20).
+		Iterator(tarantool.IterGe).
+		Key([]interface{}{uint(1010)})
+	resp, err := conn.Do(req).Get()
+	if err != nil {
+		t.Fatalf("Failed to Select: %s", err.Error())
+	}
+	if resp == nil {
+		t.Fatalf("Response is nil after Select")
+	}
+	if len(resp.Data) != expectedTuplesCount {
+		t.Fatalf("Response Data len %d != %d", len(resp.Data), expectedTuplesCount)
+	}
+}
+
+func testCrudRequestCheck(t *testing.T, req tarantool.Request,
+	resp *tarantool.Response, err error, expectedLen int) {
+	t.Helper()
+
+	if err != nil {
+		t.Fatalf("Failed to Do CRUD request: %s", err.Error())
+	}
+
+	if resp == nil {
+		t.Fatalf("Response is nil after Do CRUD request")
+	}
+
+	if len(resp.Data) < expectedLen {
+		t.Fatalf("Response Body len < %#v, actual len %#v",
+			expectedLen, len(resp.Data))
+	}
+
+	// resp.Data[0] - CRUD res.
```

Comment:
There is no in-built API to map Lua `nil, err` format to Go one? I think we should have one.

Reply:
Okay, let's create separate PR for this task.

Reply:
Is it planned to be implement right after this one or later?

Reply:
I don't understand how this API should be changed for Lua map purpose, we perform CRUD request via `Do()` and get response in `Call` format. So, how should this task affect CRUD API? As I can see the solution is a separate method like `Do(req).Get().LuaMap()`. Correct me if I misunderstood your idea.

Reply:
So there is no way to map `err` from `nil, err` to call error?

Reply:
I thought about it when I wrote the tests, it is really not comfortable for user but I couldn't find a way to hide mapping response to something like `{Res: map[string]interface{}{"metadata":..., "rows":...}, Err: ...}`. The request is executed according to the following scenario: `Do()`->`conn.send()`(returns *Future)->`putFuture()`->`pack()`(where `request.Body()` called). After `Do()` we get `future`. Then we get `response` from `future` by `Get()`.  I don't see where it would be possible to hide the mapping in this sequence of steps. That's why by now I can only offer a separate method. But if you have any ideas regarding the implementation I am open to suggestions (or maybe I just misunderstood you) .

Reply:
@oleg-jukovec what do you think?

Reply:
We could provide a custom helper type for:
https://pkg.go.dev/github.com/tarantool/go-tarantool#Future.GetTyped
with `metadata`/`rows` parsing/mapping in fields.

But we need to leave the user the opportunity to parameterize the parsing of tuples. So that the user can parse own tuple-types with a custom decoders.

Reply:
> But we need to leave the user the opportunity to parameterize parsing of tuples. So that the user can parse own tuple-types with a custom decoders.

Of course, `unflatten_rows` shouldn't be used by default. But we already receive one or two return values where each of them parsed by using the same rules as Lua function call.

Reply:
To be honest, I'm not sure that data returned by crud is always tuples -- it may be plain Lua tables.